KTL-3015 implement Kotlin LSP proxy for completions

S-furi opened the pull request:
This PR introduces a new way of performing completions requests leveraging [Kotlin-lsp](https://github.com/Kotlin/kotlin-lsp). Two approaches are included:

- **RESTful** on `api/compiler/lsp/complete`;
- **WebSocket** approach.

Currently completions with this approach supports only the latest Kotlin version, but future developments include the support of such feature by using a dedicated client and workspace for each Kotlin version supported by the kotlin-compiler-server.
Moreover, the support for different platforms than Kotlin/JVM is strictly bound to what Kotlin-lsp supports, namely only Kotlin/JVM projects as of today.

dkrasnoff commented on src/main/kotlin/com/compiler/server/controllers/LspCompletionWebSocketHandler.kt:
```diff
@@ -0,0 +1,194 @@
+        }
+        responseJobs[session.id] = responseJob
+
+        val completionWorker = scope.launch {
```
We can extract lines from 77 to 93 to a separate method called `setUpCompletionWorker` (you can think about a better name)

S-furi commented on lsp-users-projects-root/settings.gradle.kts:
```diff
@@ -0,0 +1 @@
+rootProject.name = "lsp-users-projects-root"
```
This is the base project from which LSP will retrieve base libraries/sdks/project sources. As of now, this project just sets up a simple kotlin project with version `2.2.20`, but it is thanks to this base project that we could define libraries available in playground (e.g. `kotlinx.coroutines`) or common configurations. 
The LSP needs a gradle project in its local filesystem, and I though it could be handy to have the base project here in the compiler server repo in order to reflect the current completion's approach.

dkrasnoff replied:
Hm, then it looks more like a folder in `resources`

dkrasnoff commented on src/main/kotlin/com/compiler/server/controllers/LspCompletionWebSocketHandler.kt:
```diff
@@ -0,0 +1,194 @@
+            incomingCompletionsRequests[session.id] = it
+        }
+
+        val responseFlow = MutableSharedFlow<Response>(
```
We can extract lines from 59 to 75 to a separate method called `setupResponseFlow` (you can think about a better name)

dkrasnoff commented:
Please add kdocs for all public methods and classes

dkrasnoff commented on src/main/kotlin/com/compiler/server/service/lsp/KotlinLspProxy.kt:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    fun KotlinLspProxy.onClientConnected(clientId: String) {
+        val project = Project(files = listOf(ProjectFile(name = "$clientId.kt")))
```
I'd say each project should have `clientId`, but files can then be called `file1-$clientId.kt`, `file2-$clientId.kt`, etc, so that we will have a possibility to work with mutliple files

dkrasnoff replied:
Is it hard to rewrite it to list of files instead of one file now?

dkrasnoff replied:
Got it, thanks!
Anyway, let's make multiple files here and pass to lsp only the first one. And leave a comment or todo, with the place where we choose only one file, with the explanation about LSP mode. 
Note: if it is a todo, then it should have the number of the task connected to it. If we don't have such a task, then you should create it.

dkrasnoff commented on src/main/kotlin/com/compiler/server/service/lsp/KotlinLspProxy.kt:
```diff
@@ -0,0 +1,265 @@
+    fun KotlinLspProxy.onClientConnected(clientId: String) {
+        val project = Project(files = listOf(ProjectFile(name = "$clientId.kt")))
+            .also { clientsProjects[clientId] = it }
+        val lspProject = LspProject.fromProject(project).also { lspProjects[project] = it }
```
Not clear for me, why do we have project and lspProject, can it be merged into one entity here?

dkrasnoff commented on src/main/kotlin/com/compiler/server/controllers/LspCompletionWebSocketHandler.kt:
```diff
@@ -0,0 +1,194 @@
+                    return@launch
+                }
+                lspProxy.onClientConnected(id)
+                sendResponse(Response.Init(id))
```
This method looks inside like this `outgoingResponsesFlows[id]?.tryEmit(response)`, so when `outgoingResponsesFlows[id]` is null, we don't send a message, but continue to create a worker here. Is it right, or should we `return@launch` instead?

S-furi commented on src/main/kotlin/com/compiler/server/configuration/WebSocketConfig.kt:
```diff
@@ -0,0 +1,34 @@
+
+@Configuration
+@EnableWebSocket
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
```
Yes it should really help, especially because the coexistence of Servlet and WebFlux in the same application require some tricks. I'm working on separating the completion logic to another subproject, then I'll try to migrate to WebFlux only.

S-furi commented on src/main/kotlin/com/compiler/server/controllers/LspCompletionWebSocketHandler.kt:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    private suspend fun Channel<CompletionRequest>.processIncomingRequests(session: WebSocketSession) {
+        while (true) {
```
we could actually put a condition on channel's state, like `while (!isClosedForReceive)` (*delicate coroutines api*) that should even make this function purpose clearer.

S-furi commented:
@dkrasnoff With recent commits I've made some changes that have probably changed the references to the request changes above. In short, I've
- Resolved conflicts `kotlin-community/dev` base branch
- Moved completion logic into a new springboot application module
- LSP-based completions now leverage `WebFlux`
  - Both REST endpoint (marked as suspendable) and WebSocket Implementation.
- Moved LSP related files from project root to its corresponding resources folder ([conversation reference](https://github.com/JetBrains/kotlin-compiler-server/pull/878#discussion_r2410101695))
- Applied suggested refactors

You might want to mark some conversations as resolved and maybe double-check the new WebSocket component for the refactoring suggestions you've made to the previous implementation.

If you want I can ping you when I think I've resolved everything, so that you just check once without having to check the code again.

dkrasnoff commented on completions/build.gradle.kts:
```diff
@@ -0,0 +1,27 @@
+}
+
+dependencies {
+    implementation("org.springframework.boot:spring-boot-starter-webflux")
```
lets put in `libs.versions.toml` as well

dkrasnoff commented on completions/src/main/kotlin/completions/controllers/ws/LspCompletionWebSocketHandler.kt:
```diff
@@ -0,0 +1,161 @@
+                }
+            }
+            .flatMap({ payload ->
+                val req = runCatching { objectReader.readValue<CompletionRequest>(payload) }.getOrNull()
```
Why do we need `objectReader`? Not clear for me, why just not use: `objectMapper.readValue(payload, CompletionRequest::class.java)`

dkrasnoff commented on completions/src/main/kotlin/completions/controllers/ws/LspCompletionWebSocketHandler.kt:
```diff
@@ -0,0 +1,161 @@
+    val ch: Int,
+)
+
+private val objectMapper = ObjectMapper().apply { registerKotlinModule() }
```
lets set up it in Configuration calss, and use everywhere as a bean

dkrasnoff commented on completions/src/main/kotlin/completions/lsp/LspCompletionParser.kt:
```diff
@@ -0,0 +1,142 @@
+            ?.jsonObject?.get("lookupObject")
+            ?.jsonObject?.get("lookupObject")
+
+        val importingStrategy =
```
val importingStrategy = lookupObject?.jsonObject?.get("options") ?: lookupObject

dkrasnoff commented on completions/src/main/kotlin/completions/lsp/LspCompletionParser.kt:
```diff
@@ -0,0 +1,142 @@
+                lookupObject.jsonObject["options"]
+            } else lookupObject
+
+        lookupObject?.jsonObject?.get("kind")?.let {
```
do it before getting importingStrategy

dkrasnoff commented on completions/src/main/kotlin/completions/lsp/LspCompletionParser.kt:
```diff
@@ -0,0 +1,142 @@
+    )
+}
+
+private val objectMapper = Json { ignoreUnknownKeys = true }
```
Let's make it via bean

dkrasnoff commented on completions/src/main/kotlin/completions/util/serializationUtil.kt:
```diff
@@ -0,0 +1,7 @@
+
+import com.fasterxml.jackson.databind.JsonNode
+
+object SerializationUtil {
```
the name of file starts from the lower letter
